doc/developer/design/20231204_query_lifecycle_events_logging.md

```diff
@@ -0,0 +1,149 @@
+describes how Storage sources should be made to work in an ideal
+world: after Materialize advances to a given timestamp, we should take
+note of the amount of data present in the upstream system
+corresponding to the source (for example, the max offset in Kafka),
+and only advance the source to that timestamp when all that data has
+been ingested. This is one of the parts of the currently
+semi-abandoned project called "real-time recency".
+
+I am leaving this out of scope for now as it would massively blow up
+the size of the project (from adding some relatively simple logging to
+conceptually re-architecting how Storage works). Unfortunately, this
+means that the time at which Storage sources reached a particular timestamp
+will not be very meaningful, as Storage is always allowed to just
+record things at however advanced of a timestamp it wants.
+
+However, we should still track and log how long it takes sources to
+update to the chosen timestamp, in the hope that someday we will do
+the work to make storage work this way. 
+
+### Time to result delivery
+
+I originally attempted to report the time it took to deliver the
+results over the network. However, after some thought, I realized
+there's no good way to do this. We could try to detect when all the
+bytes we've been sent have been `ACK`ed by repeatedly polling a
+Linux system call to get this information. However, this would not
+measure just the time to send the data, but also the tiem for the ACK
+to be returned to us.
+
+## Solution Proposal
+
+Create a collection called `mz_internal.mz_query_lifecycle_events`
+with the schema `(statement_execution_id uuid, event_type text,
+event_timestamp timestamp with time zone)`.
+
+The `event_type` can be one of the following:
+
+| type                                | description                                                                                                                 |
+|-------------------------------------|-----------------------------------------------------------------------------------------------------------------------------|
+| `execution_began`                   | Corresponds to today's `mz_statement_execution_history.began_at`                                                            |
+| `sources_became_ready`              | When all transitive source dependencies advanced to the timestamp. If there are none, same as `execution_began`             |
+| `compute_dependencies_became_ready` | When all compute dependencies (MVs and Indexes) advanced to the timestamp. If there are none, same as `execution_began`.    |
+| `execution_finished`                | Corresponds to today's `mz_statement_execution_history.finished_at`                                                         |
+| `last_row_was_returned`             | The time at which all rows were sent and we were ready to receive new statements from its session. |
```

Comment:
Something sounds a little funny to my ear with the variants containing "was"/"became", e.g. `sources_became_ready`.

Reply:
They are now `storage-dependencies-finished` and `compute-dependencies-finished` in the actual code.